Find a functionally equivalent code:android.media.AudioService.checkForRingerModeChange:COMMENT
Method Modifier: private     
Comment:/**
 * Checks if the adjustment should change ringer mode instead of just
 * adjusting volume. If so, this will set the proper ringer mode and volume
 * indices on the stream states.
 */

@@ -1,5 +1,5 @@
 {
-    boolean adjustVolumeIndex = true;
+    int result = FLAG_ADJUST_VOLUME;
     int ringerMode = getRingerMode();
     switch(ringerMode) {
         case RINGER_MODE_NORMAL:
@@ -11,17 +11,17 @@
                     }
                 } else {
                     // (oldIndex < step) is equivalent to (old UI index == 0)
-                    if ((oldIndex < step) && mPrevVolDirection != AudioManager.ADJUST_LOWER) {
+                    if ((oldIndex < step) && VOLUME_SETS_RINGER_MODE_SILENT && mPrevVolDirection != AudioManager.ADJUST_LOWER) {
                         ringerMode = RINGER_MODE_SILENT;
                     }
                 }
@@ -23,51 +23,55 @@
                 break;
             }
             if ((direction == AudioManager.ADJUST_LOWER)) {
-                if (mPrevVolDirection != AudioManager.ADJUST_LOWER) {
+                if (VOLUME_SETS_RINGER_MODE_SILENT && mPrevVolDirection != AudioManager.ADJUST_LOWER) {
                     ringerMode = RINGER_MODE_SILENT;
                 }
             } else if (direction == AudioManager.ADJUST_RAISE) {
                 ringerMode = RINGER_MODE_NORMAL;
             }
-            adjustVolumeIndex = false;
+            result &= ~FLAG_ADJUST_VOLUME;
             break;
         case RINGER_MODE_SILENT:
             if (direction == AudioManager.ADJUST_RAISE) {
-                if (mHasVibrator) {
-                    ringerMode = RINGER_MODE_VIBRATE;
+                if (PREVENT_VOLUME_ADJUSTMENT_IF_SILENT) {
+                    result |= AudioManager.FLAG_SHOW_SILENT_HINT;
                 } else {
-                    ringerMode = RINGER_MODE_NORMAL;
+                    if (mHasVibrator) {
+                        ringerMode = RINGER_MODE_VIBRATE;
+                    } else {
+                        ringerMode = RINGER_MODE_NORMAL;
+                    }
                 }
             }
-            adjustVolumeIndex = false;
+            result &= ~FLAG_ADJUST_VOLUME;
             break;
         default:
             Log.e(TAG, "checkForRingerModeChange() wrong ringer mode: " + ringerMode);
             break;
     }
-    setRingerMode(ringerMode);
+    setRingerMode(ringerMode, false);
     mPrevVolDirection = direction;
-    return adjustVolumeIndex;
+    return result;
 }

